tests/distributed/_test_distributed.py

```diff
@@ -0,0 +1,159 @@
+import copy
+import subprocess
+import sys
+from concurrent.futures import ThreadPoolExecutor
+from typing import Dict, List
+
+import numpy as np
+from sklearn.datasets import make_blobs, make_regression
+from sklearn.metrics import accuracy_score
+
+import lightgbm as lgb
+
+
+def create_data(task: str, n_samples: int = 1_000) -> np.ndarray:
+    """Create the appropiate data for the task.
+
+    The data is returned as a numpy array with the label as the first column.
+    """
+    if task == 'binary-classification':
+        centers = [[-4, -4], [4, 4]]
+        X, y = make_blobs(n_samples, centers=centers, random_state=42)
+    elif task == 'regression':
+        X, y = make_regression(n_samples, n_features=4, n_informative=2, random_state=42)
+    dataset = np.hstack([y.reshape(-1, 1), X])
+    return dataset
+
+
+def run_and_log(cmd: List[str]) -> None:
+    """Run `cmd` in another process and pipe its logs to this process' stdout."""
+    process = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+    assert process.stdout is not None
+
+    def stdout_stream():
+        return process.stdout.read(1)
+
+    for c in iter(stdout_stream, b''):
+        sys.stdout.buffer.write(c)
+
+
+class DistributedMockup:
+    """Simulate distributed training."""
+
+    default_config = {
+        'task': 'train',
+        'pre_partition': True,
+        'machine_list_file': 'mlist.txt',
+        'tree_learner': 'data',
+        'force_row_wise': True,
+        'verbose': 0,
+        'num_boost_round': 20,
+        'num_leaves': 15,
+        'num_threads': 2,
+    }
+
+    def __init__(self, config: Dict = {}):
+        self.config = copy.deepcopy(self.default_config)
+        self.config.update(config)
+        self.n_workers = self.config['num_machines']
+
+    def worker_train(self, i: int) -> None:
+        """Start the training process on the `i`-th worker.
+
+        If this is the first worker, its logs are piped to stdout.
+        """
+        cmd = f'./lightgbm config=train{i}.conf'.split()
+        if i == 0:
+            return run_and_log(cmd)
+        subprocess.run(cmd)
```

Comment:
Following the idea in https://github.com/microsoft/LightGBM/pull/4254#discussion_r640233072, I believe we can just call `result = subprocess.run(cmd)` and `if result.returncode != 0: raise RuntimeError`. The stderr will be there because it gets piped. What do you think?

Reply:
Yes, I guess so.